[New] EnvironmentSwitcher

tom-MO opened the pull request:
This is a new library that allows the developer to easily create a switcher for various environments. It also contains an example project demonstrating the package.

This is similar to the `EnvironmentBanner` used in the ScholarPack project. It contains some behind-the-scenes changes, but otherwise is exactly the same functionally and visually.

tom-MO committed:
[EnvironmentSwitcher] Initial commit.

tom-MO committed:
[EnvironmentSwitcher] Added code

tom-MO committed:
Removed BLoC, and other minor improvements.

BLoC was causing multiple issues with the testing (due to BuildContexts, according to error logs). As such, this was removed to accommodate a standard `setState()` approach.

tom-MO committed:
[EnvironmentSwithcer] Removed BLoC, and other minor improvements.

BLoC was causing multiple issues with the testing (due to BuildContexts, according to error logs). As such, this was removed to accommodate a standard `setState()` approach.

tom-MO committed:
Merge branch 'Environment-Switcher' of https://github.com/MyOxygen/MyOxygen-Flutter-Libraries into Environment-Switcher

tom-MO committed:
Merge branch 'Environment-Switcher' of https://github.com/MyOxygen/MyOxygen-Flutter-Libraries into Environment-Switcher

tom-MO committed:
[EnvironmentSwitcher] Small aesthetics tweak

tom-MO committed:
[EnvironmentSwitcher] Added example project.

This is easier than releasing to an existing project and testing there.

tom-MO committed:
[EnvironmentSwitcher] Prevent the FutureBuilder from calling multiple times.

tom-MO committed:
[EnvironmentSwitcher] Added test code

It's just one big test, but it runs through making sure that the banner is there, the tap handler works, and the new environment is set up.

tom-MO committed:
Merge branch 'Environment-Switcher' of https://github.com/MyOxygen/MyOxygen-Flutter-Libraries into Environment-Switcher

tom-MO committed:
[EnvironmentSwitcher] Added defaultEnvironment option.

tom-MO committed:
[EnvironmentSwitcher] Updated Readme

tom-MO committed:
[EnvironmentSwitcher] Minor tweaks to ReadMe

MO-James commented on environment_switcher/lib/src/environment.dart:
```diff
@@ -0,0 +1,31 @@
+  final String name;
+  final String description;
+  final Color bannerColor;
+  final String databaseName;
```
This is quite specific to SP, I'm not sure if it's required for a generic solution. 

Does this provide a good way to add extra fields? Subclassing might be a bit awkward.

tom-MO replied:
Good points! I've removed the `databaseName` and `value` string parameters, and replaced with a `Map<String, dynamic>` that allows adding as many additional fields as required. Would this work better?

MO-James replied:
Potentially, although it's not type-safe. And you couldn't assert that something you want to be in the map is actually in there. 

My first thoughts were making `Environment` abstract, and make `EnvironmentSwitcher` generic of the form `EnvironmentSwitcher<T extends BaseEnvironment>`. That way we can have `SPEnvironment extends BaseEnvironment` and add the `databaseName` parameter. 

Or is suppose Maybe it doesn't need to be abstract, as `Environment` will work as-is.

tom-MO replied:
Ok. In general, do we want to move away from using `Map<String, dynamic>` due it not being type-safe?

I've created an abstract class called `EnvironmentData`. This class should be extended with the necessary fields and their corresponding assertions. In the `Environment` class, there is now an optional field for `EnvironmentData`. How does that sound?

MO-James replied:
> Ok. In general, do we want to move away from using `Map<String, dynamic>` due it not being type-safe?

Yep, if a proper class can be used, it should. 

> I've created an abstract class called `EnvironmentData`. This class should be extended with the necessary fields and their corresponding assertions. In the `Environment` class, there is now an optional field for `EnvironmentData`. How does that sound?

Sounds good. Does that mean that we'd have too do: `(environment.environmentData as FooEnvironmentData).bar`?

tom-MO replied:
No, you would only have to do `environment.environmentData.bar`. I've added a load of generic types extending `EnvironmentData`, which should allow you to just simply call the data variables without having to typecast.

MO-James commented on environment_switcher/lib/src/environment.dart:
```diff
@@ -0,0 +1,31 @@
+
+  bool get isNameValid => name != null && name.trim().isNotEmpty;
+
+  Environment({
```
What do each of the parameters mean?

MO-James commented on environment_switcher/lib/src/environment_store.dart:
```diff
@@ -0,0 +1,30 @@
+    if (environment == null) {
+      await store.setString(null, key: _environmentKey);
+    } else {
+      await store.setString(environment.name, key: _environmentKey);
```
Could this just be a one-liner `store.setString(environment?.name, key: _environmentKey);`.
Would `value` make more sense than `name`?

MO-James commented on environment_switcher/lib/src/environment_switcher.dart:
```diff
@@ -0,0 +1,195 @@
+export 'environment_store.dart';
+
+class EnvironmentSwitcher extends StatefulWidget {
+  final Widget Function(Environment) builder;
```
`BuildContext` may also be required for a generic solution

tom-MO replied:
Why is that? Wouldn't the `BuildContext` from the `build()` method be fine?

MO-James replied:
Because it's using a context from higher up the tree. Take for example:

```
Widget build(BuildContext contextA) {
  return BlocProvider<FooBloc>(
    bloc: _fooBloc,
    child: EnvironmentSwitcher(
      builder: (env) {
        final fooBloc = BlocProvider.of<FooBloc>(contextA);
        return MyWidget(env: env);
      }),
    )
  );
}
```

Which would throw because `contextA` doesn't contain a `BlocProvider`.

It's not required where it's used in SP, but could lead to unexpected errors in future apps. This is why builder functions typically pass back a context.

tom-MO replied:
Ah yes! that makes sense. I misunderstood where you were referring to put the `BuildContext`. I'll do that now.

MO-James commented on environment_switcher/lib/src/environment_switcher.dart:
```diff
@@ -0,0 +1,195 @@
+  /// A banner that visually shows the user what [Environment]
+  /// is currently set to.
+  /// This is useful because it's possible to be on one Flavour
+  /// in the native code, and another in the Flutter code.
```
This comment is actually wrong now. We're not setting different icons/app titles ect for environments, so they're entirely flutter-side.

MO-James commented on environment_switcher/lib/src/environment_switcher.dart:
```diff
@@ -0,0 +1,195 @@
+  }
+
+  _StateEnvironmentSwitcher(this.environments, EnvironmentStore store)
+      : environmentStore = store ?? EnvironmentStore(store: Store());
```
Why pass these in the constructor and not use `widget.foo`?

MO-James commented on environment_switcher/lib/src/environment_switcher.dart:
```diff
@@ -0,0 +1,195 @@
+  const EnvironmentSwitcher({
+    @required this.builder,
+    @required this.environments,
+    this.environmentStore, // mainly required for testing
```
Why? How is it used?

tom-MO replied:
I'm not entirely sure why, but I assume it's because the Store library tries to access native code, which the testing platform doesn't seem to like. You can try it out by commenting out the `_MockEnvironmentStore` instance in the test. It throws a testing error. Creating a mock environment store seems to avoid this issue.

MO-James replied:
Oh I see! It creates a default version otherwise, and that default uses platform code so can't run in unit tests.

MO-James commented on environment_switcher/lib/src/environment_switcher.dart:
```diff
@@ -0,0 +1,195 @@
+        throw "No environment saved";
+      }
+      final environment =
+          environments.firstWhere((env) => env.name == savedEnvironmentName, orElse: () => null);
```
Is matching on name correct? Are we doing any asserts to check that name is always unique? Is value more appropriate?

tom-MO replied:
I agree this is a bit questionable. I reckoned it's highly unlikely (also bad practice) that two environments would have the same name. Short of creating a way of converting the object to string and back (or use `JsonSerialization`), this would have to do.

MO-James commented on environment_switcher/lib/src/environment_switcher.dart:
```diff
@@ -0,0 +1,195 @@
+          environments.firstWhere((env) => env.name == savedEnvironmentName, orElse: () => null);
+      if (environment == null) {
+        throw "Environment not found in list";
+      }
```
This could go in the `orElse` block to be slightly clearer IMO.

MO-James commented on environment_switcher/lib/src/environment_switcher.dart:
```diff
@@ -0,0 +1,195 @@
+      return environment;
+    } catch (e) {
+      print(e.toString());
+    }
```
What do you expect to throw? Is it just the "no environment saved"? If so, it might be better to just use an `if/else`.

MO-James commented on environment_switcher/lib/src/environment_switcher.dart:
```diff
@@ -0,0 +1,195 @@
+  final List<Environment> environments;
+  final EnvironmentStore environmentStore;
+
+  Environment currentEnvironment;
```
Is there any way for a child widget to get the current environment. If not, maybe use an `InheritedWidget` now that we can't use `BlocProvider`

MO-James commented on environment_switcher/lib/src/select_environment_sheet.dart:
```diff
@@ -0,0 +1,98 @@
+  final Environment currentEnvironment;
+  final void Function(Environment) onNewEnvironmentSelected;
+
+  SelectEnvironmentSheet({
```
Can this be `const`?

MO-James commented on environment_switcher/lib/src/select_environment_sheet.dart:
```diff
@@ -0,0 +1,98 @@
+              "mock data. It'll restart the app in order to make sure all the "
+              "data is fresh.",
+            ),
+          ),
```
Might be good for this text to be overridable.

MO-James commented on environment_switcher/lib/src/select_environment_sheet.dart:
```diff
@@ -0,0 +1,98 @@
+
+  @override
+  Widget build(BuildContext context) {
+    final subtitle = "${type.description}" + (type.value == null ? "" : "\n${type.value}");
```
Split into multiple lines please.

MO-James reviewed (commented):
Some things here work for SP but won't apply well for a generic solution. 

Doesn't look like any sensitive strings from SP were committed 👍

Have you got a PR for how this would look within SP?

tom-MO committed:
Replaced the string parameters databaseName and value with Map parameter values.

The parameters `databaseName` and `value` are currently specific to a project, which means this isn't easily transferable to another project. It also means there is no ability to add additional fields/values/information to the Environment object.

These two parameters have been replaced with a Map<String, dynamic> object named `values`. This Map object should contain any additional data required for the Environment object (different URL, mock/live server, fake/real data etc). Using a Map allows the developer to add as much additional data as required.

tom-MO committed:
Replaced Map with custom EnvData class.

A custom abstract class (extendable) allows the developer to add all necessary parameters/fields and associated assertions. It is an optional field for `Environment`.

tom-MO committed:
Single-lined Store code.

tom-MO committed:
Removed incorrect comment.

tom-MO committed:
Removed state constructor, as it wasn't needed.

The `_StateEnvironmentSwitcher` constructor is not needed. Can't remember why I included it in the first place. Maybe because I wanted to keep the `EnvironmentSwitcher` const constructor?

Following from the change above, the `builder` field was changed to `childBuilder` as that is more descriptive.

tom-MO committed:
Updated test with required field

Also added some text to better explain an assertion.

tom-MO committed:
Added BuildContext to the childBuilter

This should prevent calling the parent context when building the child. Calling the parent context may not yield the expected results.

tom-MO committed:
Removed unnecessary try/catch block

tom-MO committed:
Made SelectEnvironmentSheet constructor const.

tom-MO committed:
Allow overriding selection sheet text.

tom-MO committed:
Added inherited widget to access the current environment from further down the tree.

tom-MO committed:
Added generic type to pass the additional environment data.

This also fixes issues with calling the inherited widget to get the current environment.

tom-MO committed:
Updated CircleCI test check.

tom-MO commented:
@MO-James Addressed your comments.
 
> Have you got a PR for how this would look within SP?

I haven't done that, but I have included a test project to show the library in use. Is it also worth considering implementing this in SAM?

MO-James commented:
> @MO-James Addressed your comments.
> 
> > Have you got a PR for how this would look within SP?
> 
> I haven't done that, but I have included a test project to show the library in use. Is it also worth considering implementing this in SAM?

I was thinking in SP as a regression test.

MO-James commented on environment_switcher/example/lib/page.dart:
```diff
@@ -0,0 +1,162 @@
+  const EnvData(this.someString, this.someValue);
+
+  @override
+  List<Object> get props => null;
```
Should this be overridden?

MO-James commented on environment_switcher/example/lib/page.dart:
```diff
@@ -0,0 +1,162 @@
+  // used by the build method of the State. Fields in a Widget subclass are
+  // always marked "final".
+
+  final String title;
```
[Style] variables above constructor please.

MO-James commented on environment_switcher/example/lib/page.dart:
```diff
@@ -0,0 +1,162 @@
+      // called again, and so nothing would appear to happen.
+      _counter++;
+    });
+  }
```
Do we need all this counter stuff in the example app?

MO-James commented on environment_switcher/example/lib/page.dart:
```diff
@@ -0,0 +1,162 @@
+        final environment = EnvironmentSwitcher.of<EnvData>(ctx)?.currentEnvironment;
+        print(
+            "EnvironmentSwitcher context at MyHomePage builder: ${EnvironmentSwitcher.of<EnvData>(ctx)}");
+
```
Is this still required?

MO-James commented on environment_switcher/example/lib/page.dart:
```diff
@@ -0,0 +1,162 @@
+          // how it positions its children. Here we use mainAxisAlignment to
+          // center the children vertically; the main axis here is the vertical
+          // axis because Columns are vertical (the cross axis would be
+          // horizontal).
```
All of this default example stuff distracts from the lib

MO-James commented on environment_switcher/lib/src/environment_switcher.dart:
```diff
@@ -7,118 +8,162 @@ import 'select_environment_sheet.dart';
+class InheritedEnvironmentSwitcher<T extends EnvironmentData> extends InheritedWidget {
+  final Environment<T> currentEnvironment;
+
+  InheritedEnvironmentSwitcher({
```
Can this class be private?

MO-James commented on environment_switcher/lib/src/environment_switcher.dart:
```diff
@@ -128,32 +173,31 @@ class _StateEnvironmentSwitcher extends State<EnvironmentSwitcher> {
-class _Banner extends StatelessWidget {
-  final Environment environment;
-  final Widget child;
+class LocalBanner<T extends EnvironmentData> extends StatelessWidget {
```
Can this class be private?

MO-James commented on environment_switcher/test/environment_switcher_test.dart:
```diff
@@ -9,37 +9,32 @@ final environments = [
+          child: Container(
+            width: 100,
+            height: 100,
+            //color: Colors.black,
```
Commented out code

MO-James reviewed (commented):
Code review. Need to run the sample app.

MO-James commented:
@thomas-bartleet  Just noticed your commits are missing the [EnvironmentSwitcher] tag

tom-MO committed:
[EnvironmentSwitcher] General fixes

Changes:
- Removed unnecessary example and commented code.
- Added variables to `props`.

tom-MO commented:
> @thomas-bartleet Just noticed your commits are missing the [EnvironmentSwitcher] tag

Oh fuuuuuuuu... dge-cake.

tom-MO commented:
This library is causing some issues with the ScholarPack project, but I don't think it's to do with the actual library itself. I think it has more to do with the way the code is set up and how Flutter operates. I will try to explain (so that there is a written record and I don't forget!).

The `EnvironmentSwitcher` requires a `Navigator` instance to open the bottom modal sheet, which `MaterialApp` internally creates. This means `EnvironmentSwitcher` needs to be **below** `MaterialApp` in the widget tree. In addition, `EnvironmentSwitcher` needs a new instance of the `BuildContext` to access the `Navigator` instance (as well as `MediaQuery` and `MaterialLocalization` instances). This can be sorted easily by putting a `Builder` widget between them:

```dart
MaterialApp(
  // Using the `home` property as an example. This also applies to using the "generated routes".
  home: Builder(
    builder: (context) => EnvironmentSwitcher(
        environments: ...
        defaultEnvironment: ...
        childBuilder: (cntxt) => MyHomePage(),
      ),
  ),
);
```

This approach works fine for smaller apps, with very little (if any) structure/hierarchy/architecture. However, with larger apps that have various dependencies and repositories, this simply won't be enough. Take the example below:

```dart
MaterialApp(
  home: Builder(
    builder: (context) => EnvironmentSwitcher(
        environments: ...
        defaultEnvironment: ...
        childBuilder: (cntxt) => MyHomePage(),
      ),
  ),
  builder: (_, child) => DependencyInjection(child: child),
);
```

Let's say my `builder` property simply returns an instance of `DependencyInjection` and passes the child to this instance. At what point in the tree hierarchy does the `DependencyInjection` widget get placed? [According to the docs](https://api.flutter.dev/flutter/material/MaterialApp/builder.html), the result from the `builder` property gets placed after the `MediaQuery` and `MaterialLocalization` instances are created, but **before** the `Navigator` widget! So, because it is before the `Navigator` widget, it is placed higher up the tree than the `Builder` widget. With the above code, the (simplified) widget tree now looks like:

```
MaterialApp
|- MediaQuery
|-- MaterialLocalization
|--- DependencyInjection
|---- Navigator
|----- Builder
|------ EnvironmentSwitcher
|------- MyHomePage
```

To change the dependencies using the `EnvironmentSwitcher`, we have two options (AFAICS):
1.  The `EnvironmentSwitcher` needs to be above the `DependencyInjection` widget. However, we need access to `Navigator`. So this doesn't work.
2. Move the `DependencyInjection` to below the `EnvironmentSwitcher`. However, this means that `DependencyInjection` gets a new instance for every page. So this doesn't work.

Because of the above, we would have to restructure the main code to fit this library, and I don't think that's right. On potential solution might be to look at implementing a custom bottom modal sheet that doesn't use the `Navigator` instance (instead it uses a `Stack`, a `Container`m, and some animation, for example), which should allow the `EnvironmentSwitcher` to be above the `DependencyInjection` widget in the tree above.

Due to time constraints and other priorities, the aforementioned suggestion will not be implemented at this time. Therefore, I'll close this PR until time is available and/or I'm twiddling my thumbs for something to do.

MO-James commented:
@thomas-bartleet 
That makes sense. How is SP currently structured? As this code originally came from SP and it's working there.

tom-MO commented:
In SP, the `EnvironmentBanner` uses a BLoC architecture such that the BLoC is above the entire app. However, the actual `EnvironmentBanner` widget is at the `home` property level. The widget tree would look like:

```
EnvironmentBannerBloc
|- MaterialApp
|-- MediaQuery
|--- MaterialLocalization
|---- DependencyInjection
|----- Navigator
|------ EnvironmentBanner
|------- MyHomePage
```
